### PR TITLE
fix: add akv support to the module

### PR DESCRIPTION
**Pull request by ghost**
I have upgraded the azure key vault terraform code to version "3.0.2"
Please check and review the changes that I have made
This should work I guess
Proof of the terraform code working is attached in the below image
please have a look at it
image

### CODE REVIEW

File: terraform-jx-azurekeyvault/main.tf [hunk 1]
```diff
@@ -46,12 +51,12 @@ resource "azurerm_key_vault" "jx" {
 resource "azurerm_key_vault_access_policy" "jx" {
   count        = var.enabled ? 1 : 0
   key_vault_id = azurerm_key_vault.jx.0.id
-  object_id    = var.principal_id
+  object_id    = data.azurerm_client_config.current.object_id
```
Comment (ankitm123) [inline terraform-jx-azurekeyvault/main.tf:54]: Why this change? Wont it break existing deployments? I dont know azure, but a data source will query something that already exists, so is this something azure already does by default?

File: terraform-jx-azurekeyvault/main.tf [hunk 2]
```diff
@@ -6,11 +6,16 @@ terraform {
   required_version = ">= 0.13.2"
   required_providers {
     azurerm = {
-      version = ">=2.48.0, <= 2.61.0"
+      version = "~>3.0.1"
     }
   }
 }
 
+provider "azurerm" {
```
Comment (ankitm123) [inline terraform-jx-azurekeyvault/main.tf:14]: Why do we need to define this again here? We define it at the top level main.tf file ...

File: terraform-jx-azurekeyvault/variables.tf [hunk 3]
```diff
@@ -25,10 +25,10 @@ variable "key_vault_sku" {
   description = "SKU of the Key Vault resource to create. Valid values are standard or premium"
   default     = "standard"
 }
-variable "principal_id" {
-  type        = string
-  description = "The id of the service principal that should be granted permission on the key vault"
-}
+# variable "principal_id" {
```
Comment (ankitm123) [inline terraform-jx-azurekeyvault/variables.tf:28]: If we are removing this, we can just remove, and not comment ...

**Review by ankitm123 — COMMENTED** [review]
Some generic comments ...

File: terraform-jx-azurekeyvault/main.tf [hunk 4]
```diff
@@ -6,11 +6,16 @@ terraform {
   required_version = ">= 0.13.2"
   required_providers {
     azurerm = {
-      version = ">=2.48.0, <= 2.61.0"
+      version = "~>3.0.1"
```
Comment (ankitm123) [inline terraform-jx-azurekeyvault/main.tf:9]: Will it break if we do >=2.48.0 < 4.0.0? Is there a way to preserve backward compatibility, ppl using tf lock file have to run terraform init -upgrade, which may be ok :thinking: -> the test I would do is create a cluster with the code in main, and then upgrade the module to ur version, and do a plan/apply, and see if that leads to weird issues ...

**Comment by ankitm123** [comment]
Change the commit message and the PR title to follow conventional commit -> so something like:
```
fix: add akv support to the module
```

In case of commit, if this is a breaking change, remember to add BREAKING CHANGE: in commit footer.
https://jenkins-x.io/community/code/#the-commit-message

File: terraform-jx-azurekeyvault/main.tf (hunk 1 above)
Comment (avinashnadendla) [reply terraform-jx-azurekeyvault/main.tf:54]: Actually it's an object id
We are already using a data source for it from before itself
So I followed the official docs it is said that it's better to use a data source for it rather than a variable
So probably won't break anything

File: terraform-jx-azurekeyvault/main.tf (hunk 2 above)
Comment (avinashnadendla) [reply terraform-jx-azurekeyvault/main.tf:14]: Actually it's because of the update in terraform provider
We have to define a provider to make it work

File: terraform-jx-azurekeyvault/main.tf (hunk 4 above)
Comment (avinashnadendla) [reply terraform-jx-azurekeyvault/main.tf:9]: Actually the previous versions has a different kind of "get" permissions
Like "get" has been changed to "Get"
So i don't think so whether it will work for that

Comment (ankitm123) [reply terraform-jx-azurekeyvault/main.tf:9]: Didn't that get fixed in version 2.61.X? Does it also not work with that version?

File: terraform-jx-azurekeyvault/main.tf (hunk 2 above)
Comment (ankitm123) [reply terraform-jx-azurekeyvault/main.tf:14]: defining multiple providers sounds a bit weird to me, do you have links to the docs for this?

File: terraform-jx-azurekeyvault/main.tf (hunk 1 above)
Comment (ankitm123) [reply terraform-jx-azurekeyvault/main.tf:54]: Ah I see, it's defined here: https://github.com/jenkins-x-terraform/terraform-jx-azure/blob/1071a3749773c502740820ca17907d9bf170aa2a/terraform-jx-azurekeyvault/main.tf#L20 
Can you point me to the docs for this? I see this: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault_access_policy#object_id
If we change it, then it will lead to re-creation?

Also, curious, why this was provided as a variable, do you know? Will it be safe to check if the variable is empty, then use the data source? `var.principle_id != "" ? var.principle_id : data.azurerm_client_config.current.object_id`

File: terraform-jx-azurekeyvault/main.tf (hunk 4 above)
Comment (avinashnadendla) [reply terraform-jx-azurekeyvault/main.tf:9]: Actually, it's working for 2.61.X version but it ain't working for any version >3.0.0

**Review by tomhobson — CHANGES_REQUESTED** [review]
Re-request when @ankitm123's comments have been addressed,

Get in touch with me on slack if you need any additional help.
